src/script/descriptor.cpp

```diff
@@ -1143,36 +1143,46 @@ std::unique_ptr<DescriptorImpl> ParseScript(uint32_t& key_exp_index, Span<const
         ++key_exp_index;
         std::vector<std::unique_ptr<DescriptorImpl>> subscripts;
         std::vector<int> depths;
-        while (expr.size()) {
+        if (expr.size()) {
+            std::vector<bool> counts; //!< whether we already had an expression at that depth
             if (!Const(",", expr)) {
                 error = strprintf("tr: expected ',', got '%c'", expr[0]);
                 return nullptr;
             }
-            auto sarg = Expr(expr);
-            auto colon = std::find(sarg.begin(), sarg.end(), ':');
-            if (colon == sarg.end()) {
-                error = "tr() requires scripts with 'depth:' prefix";
-                return nullptr;
-            }
-            uint32_t depth;
-            if (!ParseUInt32(std::string(sarg.begin(), colon), &depth)) {
-                error = "depth is not numeric in tr()";
-                return nullptr;
-            }
-            if (depth > TAPROOT_CONTROL_MAX_NODE_COUNT) {
-                error = strprintf("depth is out of range in tr() (0 <= depth <= %i)", TAPROOT_CONTROL_MAX_NODE_COUNT);
-                return nullptr;
+            while (expr.size()) {
+                if (Const("{", expr)) {
+                    counts.push_back(false);
+                    if (counts.size() > TAPROOT_CONTROL_MAX_NODE_COUNT) {
+                        error = strprintf("tr() supports at most %i nesting levels", TAPROOT_CONTROL_MAX_NODE_COUNT);
+                        return nullptr;
+                    }
+                    continue;
+                }
+                auto sarg = Expr(expr);
+                auto subdesc = ParseScript(key_exp_index, sarg, ParseScriptContext::P2TR, out, error);
+                if (!subdesc) return nullptr;
+                depths.push_back(counts.size());
+                subscripts.push_back(std::move(subdesc));
+                while (counts.size() && counts.back()) {
+                    if (!Const("}", expr)) {
+                        error = strprintf("tr(): expected '}' after script expression");
+                        return nullptr;
+                    }
+                    counts.pop_back();
+                }
+                if (counts.size()) {
```

Comment:
Behavior Note:

I believe this would be more robust to go before the while loop -- it shouldn't come up, because we're in a binary tree maybe -- but it would be better to explicitly check:
```c++
if (counts.size() && counts.back() == false) 
```
or something like
```c++
if (counts.size() && counts.back() <= 2)
```

if we want to limit how many terms in a nest are allowed.

right now this invariant is implicitly checked by the order of the code and it is non-obvious to me.

Reply:
I've restructured things a bit, and added a bunch of comments. Does this look better?